pkg/db/transaction.go

```diff
@@ -5,30 +5,65 @@ import (
 	"fmt"
 	"github.com/pkg/errors"
 	"go.uber.org/zap"
+	"math/rand"
 	"sync"
+	"time"
 )
 
+const txMaxRetries = 3
+const txMaxJitter = 350
+const txMinJitter = 25
+
 func TransactionResult(conn Connection, dbOps func(tx *TxConnection) (interface{}, error), logger *zap.SugaredLogger) (interface{}, error) {
+	var result interface{}
+	var txErr error
+
 	log := func(msg string, args ...interface{}) {
-		if logger != nil {
-			logger.Debugf(msg, args...)
+		if logger == nil {
+			return
 		}
+		logger.Infof(msg, args...)
 	}
-	txConnection, err := conn.Begin()
-	if err != nil {
-		return nil, err
-	}
 
-	result, err := dbOps(txConnection)
-	if err != nil {
-		log("Rollback transactional DB context because an error occurred: %s", err)
-		if rollbackErr := txConnection.Rollback(); rollbackErr != nil {
-			err = errors.Wrap(err, fmt.Sprintf("Rollback of db operations failed: %s", rollbackErr))
+	for i := 0; i < txMaxRetries; i++ {
+		if i > 0 {
```

Comment:
Can we add some commit to explain this behavior? is the purpose here to give some delay to prevent concurrent transaction calls (like the previous one is during rollback and prevent new one begin)?

Reply:
Sure, will add a comment.

What is does is: if the commit failes caused by collsiions, it will rollback the transaction and run it from scratch again until `txMaxRetries` is reached